### PR TITLE
module: bwa/sampe

### LINKED ISSUE #589: new module: BWA/ALN
I think it would be good to have a module for [BWA/ALN](https://bioconda.github.io/recipes/bwa/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### LINKED ISSUE #622: new module: BWA/SAMPE
I think it would be good to have a module for [BWA SAMPE](https://bioconda.github.io/recipes/bwa/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by jfy133**
Adds `bwa sampe`, requiring #624  

## PR checklist

Closes #589  and Closes #622 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/bwa/sampe/meta.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      description: FASTQ files
+      pattern: "*.{fastq,fq}.gz"
+  - sai:
+      type: file
+      description: SAI file
+      pattern: "*.sai"
+  - index:
+        type: file
+        description: BWA genome index files
+        pattern: "Directory containing BWA index *.{amb,ann,bwt,pac,sa}"
```
Comment (grst) [inline modules/bwa/sampe/meta.yml:40]: I understand that it should be a directory containing those 5 files. 
However `pattern` should be a valid glob pattern, and the given pattern matches *any* of those files, not all of them. 

With `STAR` we just used `star` as pattern: 
https://github.com/nf-core/modules/blob/2a7c60e9652fc37ad39e784f1b87b218fae8c94c/modules/star/align/meta.yml#L30

Not sure what's best.

File: modules/bwa/sampe/main.nf [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    tuple val(meta), path("*.bam"), emit: bam
+    path "*.version.txt"          , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    def read_group = meta.read_group ? "-r ${meta.read_group}" : ""
+
+    """
+    INDEX=`find -L ./ -name "*.amb" | sed 's/.amb//'`
```
Comment (grst) [inline modules/bwa/sampe/main.nf:35]: is this necessary? Shouldn't `$index` point to that directory anyway?

File: modules/bwa/sampe/meta.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
+  - bam:
+      type: file
+      description: BAM/CRAM/SAM file
+      pattern: "*.{bam,cram,sam}"
```
Comment (grst) [inline modules/bwa/sampe/meta.yml:55]: According to the output channel, this will only emit a BAM file

**Review by grst — COMMENTED** [review]
A few small things (documenation mostly), other than that LGTM

File: modules/bwa/sampe/meta.yml (hunk 1 above)
Comment (jfy133) [reply modules/bwa/sampe/meta.yml:40]: Oops, that was a copy-paste error for the pattern.

But re the variable: I also thought that, but I had to follow what was done in [`BOWTIE2`](https://github.com/nf-core/modules/blob/2a7c60e9652fc37ad39e784f1b87b218fae8c94c/modules/bowtie2/align/main.nf#L37-L48) as reference (which works).

When I just supply `$index` it gets an emitted variable of `./bwa/genome` (which is the directory and reference prefix), I'm not sure why tbh. The `$INDEX` gubbins is to get just the reference directory i.e. `bwa/` from the `bwa index` module that `bwa sampe` accepts.

I've updated the pattern (somewhat) acordingly.

File: modules/bwa/sampe/main.nf (hunk 2 above)
Comment (jfy133) [reply modules/bwa/sampe/main.nf:35]: See comment above: https://github.com/nf-core/modules/pull/625#discussion_r680458792

The way `BWA_INDEX` outputs stuff made this more complicated, and I followed `BOWTIE2_ALIGN` conventions

Comment (grst) [reply modules/bwa/sampe/main.nf:35]: Ok, great! Thanks for explaining!

**Comment by grst** [comment]
Let's just wait for additional input on https://github.com/nf-core/modules/pull/626#issuecomment-890461278, other than that this LGTM!

**Review by grst — APPROVED** [review]
LGTM

**Pull request merged**